src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs

```diff
@@ -0,0 +1,265 @@
+                RegexOptions.ECMAScript |
+                RegexOptions.CultureInvariant;
+            if ((regexMethod.Options.Value & ~(int)SupportedOptions) != 0)
+            {
+                return Diagnostic.Create(DiagnosticDescriptors.InvalidRegexArguments, methodSyntax.GetLocation(), "options");
+            }
+
+            // Validate the timeout
+            if (regexMethod.MatchTimeout.Value is 0 or < -1)
+            {
+                return Diagnostic.Create(DiagnosticDescriptors.InvalidRegexArguments, methodSyntax.GetLocation(), "matchTimeout");
+            }
+
+            // Parse the input pattern
+            try
+            {
+                regexMethod.Tree = RegexParser.Parse(regexMethod.Pattern, (RegexOptions)regexMethod.Options, culture);
+                regexMethod.Code = RegexWriter.Write(regexMethod.Tree);
+            }
+            catch (Exception e)
+            {
+                return Diagnostic.Create(DiagnosticDescriptors.InvalidRegexArguments, methodSyntax.GetLocation(), e.Message);
+            }
+
+            // Determine the namespace the class is declared in, if any
+            string? ns = regexMethodSymbol?.ContainingType?.ContainingNamespace?.ToDisplayString(
+                SymbolDisplayFormat.FullyQualifiedFormat.WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted));
+
+            var rc = new RegexType
+            {
+                Keyword = typeDec is RecordDeclarationSyntax rds ? $"{typeDec.Keyword.ValueText} {rds.ClassOrStructKeyword}" : typeDec.Keyword.ValueText,
+                Namespace = ns,
+                Name = $"{typeDec.Identifier}{typeDec.TypeParameterList}",
+                Constraints = typeDec.ConstraintClauses.ToString(),
+                ParentClass = null,
+                Method = regexMethod,
+            };
+
+            RegexType current = rc;
+            var parent = typeDec.Parent as TypeDeclarationSyntax;
+
+            while (parent is not null && IsAllowedKind(parent.Kind()))
+            {
+                current.ParentClass = new RegexType
+                {
+                    Keyword = parent is RecordDeclarationSyntax rds2 ? $"{parent.Keyword.ValueText} {rds2.ClassOrStructKeyword}" : parent.Keyword.ValueText,
+                    Namespace = ns,
+                    Name = $"{parent.Identifier}{parent.TypeParameterList}",
+                    Constraints = parent.ConstraintClauses.ToString(),
+                    ParentClass = null,
+                };
+
+                current = current.ParentClass;
+                parent = parent.Parent as TypeDeclarationSyntax;
+            }
+
+            return rc;
+
+            static bool IsAllowedKind(SyntaxKind kind) =>
+                kind == SyntaxKind.ClassDeclaration ||
+                kind == SyntaxKind.StructDeclaration ||
+                kind == SyntaxKind.RecordDeclaration ||
+                kind == SyntaxKind.RecordStructDeclaration;
```

Comment:
Why not interfaces? This should be allowed unless I'm missing something

```csharp
public partial interface C
{
    [RegexGenerator]
    partial void M();
}

public partial interface C
{
    partial void M() { }
}
```

Reply:
Because... it didn't occur to me anyone would want to do this :)  But, I've added it.